File: 0767-reorganize-string/0767-reorganize-string.py
class Solution:
    def reorganizeString(self, s: str) -> str:
        ans, counter, pq = [], dict(), []
        for c in s:
            counter[c] = counter.get(c, 0) + 1
        
        for k, count in counter.items():
            heappush(pq, (-count, k))
        
        while pq:
            count_first, char_first = heappop(pq)
            
            if not ans or char_first != ans[-1]:
                ans.append(char_first)
                if count_first + 1 != 0:
                    heappush(pq, (count_first + 1, char_first))
            else:
                if not pq:
                    return ""
                count_second, char_second = heappop(pq)
                ans.append(char_second)
                if count_second + 1 != 0:
                    heappush(pq, (count_second + 1, char_second))
                heappush(pq, (count_first, char_first))
            
        return "".join(ans)

File: 0767-reorganize-string/0767-reorganize-string.cpp
class Solution {
public:
    string reorganizeString(string s) {
        vector<int> alphabets(26);
        for (char& c : s) ++alphabets[c - 'a'];
        
        int maxCount = 0, letter = 0, index = 0;
        for (int i = 0; i < 26; ++i){
            if (alphabets[i] > maxCount){
                maxCount = alphabets[i];
                letter = i;
            }
        }
        
        if (maxCount > (s.size() + 1) / 2) return "";
        
        string ans = s;
        while (alphabets[letter] != 0){
            ans[index] = char(letter + 'a');
            index += 2;
            --alphabets[letter];
        }
        
        for (int i = 0; i < 26; ++i){
            while (alphabets[i] != 0){
                if (index >= s.size()) index = 1;
                ans[index] = char(i + 'a');
                index += 2;
                --alphabets[i];
            }
        }
        
        return ans;
    }
};